src/core/BigDecimal.ts

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * A wrapper for BigDecimal values. You can use this class to store and query BigDecimal values in distributed objects.
+ *
+ * A `BigDecimal` consists of an arbitrary precision number {@link unscaledValue} and a {@link scale}. If zero or positive,
+ * the scale is the number of digits to the right of the decimal
+ * point. If negative, the unscaled value of the number is multiplied
+ * by ten to the power of the negation of the scale.  The value of the
+ * number represented by the `BigDecimal` is therefore
+ * <code>(unscaledValue &times; 10<sup>-scale</sup>)</code>.
+ */
+export class BigDecimal {
+
+    /** @internal */
+    private static readonly MAX_EXPONENT_DIGITS = Number.MAX_SAFE_INTEGER.toString().length;
+
+    /**
+     * Unscaled value of this `BigDecimal`. This value is a native JavaScript BigInt object, which is used to store
+     * digits of this BigDecimal.
+     */
+    readonly unscaledValue: BigInt;
+
+    /**
+     * Scale of this `BigDecimal`.  If zero or positive, the scale
+     * is the number of digits to the right of the decimal point.
+     * If negative, the unscaled value of the number is multiplied
+     * by ten to the power of the negation of the scale. For example,
+     * a scale of `-3` means the unscaled value is multiplied by 1000.
+     */
+    readonly scale: number;
+
+    /** @internal */
+    private constructor(unscaledValue: BigInt, scale: number) {
+        this.unscaledValue = unscaledValue;
+        this.scale = scale;
+    }
+
+    /** @internal */
+    private static parseExp(input: string, offset: number, len: number): number {
+        let exp = 0;
```

Comment:
parseExp parses a exponent string like e+23, e-23, e23 and checks for invalid ones. TBH I do not trust js parseInt behaviour. To keep the consistency of jdk implementation inspiration, and since I tested this method 

For example 

`parseInt('1232312322222.2')` returns `1232312322222` but I want it to throw since it is not a valid integer.